File: java/CombinationSum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (candidates == null || candidates.length == 0){
            return res;
        }
        List<Integer> element = new ArrayList<Integer>();
        dfs(res, element, candidates, target, 0, 0);
        return res;        
    }
    
    private void dfs(List<List<Integer>> res, List<Integer> element, int[] candidates, int target, int index, int sum){
        if (sum > target){
            return;
        }
        
        if (sum == target){
            List<Integer> newElement = new ArrayList<Integer>(element);
            res.add(newElement);
            return;
        }
        
        for (int i = index; i < candidates.length; i++){
            element.add(candidates[i]);
            List<Integer> newElement = new ArrayList<Integer>(element);
            dfs(res, element, candidates, target, i, sum + candidates[i]);
            element.remove(element.size() - 1);
        }
        
    }
}

File: python/CombinationSum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        self.dfs(res, [], candidates, target, 0)
        return res
    
    def dfs(self, res, path, candidates, target, index):
        if target < 0:
            return
        if target == 0:
            res.append(path)
            return
        for i in range(index, len(candidates)):
            self.dfs(res, path + [candidates[i]], candidates, target - candidates[i], i)
